primefaces-selenium/primefaces-selenium-components/src/main/java/org/primefaces/selenium/component/FileUpload.java

```diff
@@ -0,0 +1,139 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.primefaces.selenium.component.base.AbstractInputComponent;
+import org.primefaces.selenium.component.base.ComponentUtils;
+import org.primefaces.selenium.internal.ConfigProvider;
+import org.primefaces.selenium.spi.WebDriverProvider;
+
+/**
+ * Component wrapper for the PrimeFaces {@code p:fileUpload}.
+ */
+public abstract class FileUpload extends AbstractInputComponent {
+
+    @Override
+    public WebElement getInput() {
+        // in case of mode=simple skinSimple=false the input element is this element
+        boolean isInputFile = "input".equals(getTagName()) && "file".equals(getAttribute("type"));
+        return isInputFile ? this : findElement(By.id(getId() + "_input"));
+    }
+
+    /**
+     * Returns the input´s value.
+     * @return the file name
+     */
+    public String getValue() {
+        return getInput().getAttribute("value");
+    }
+
+    /**
+     * Sets the input's value.
+     * This should be the files absolute path.
+     * @param value the file name to set
+     */
+    public void setValue(Serializable value) {
+        if (getInput() != this) {
+            // input file cannot be cleared if skinSimple=false
+            getInput().clear();
+        }
+        ComponentUtils.sendKeys(getInput(), value.toString());
+    }
+
+    /**
+     * Sets the input's value from given files.
+     * @param values the file name(s) to set
+     */
+    public void setValue(File... values) {
+        // several references in the WEB state that (at least) Firefox and Chrome accept
+        // multiple filenames separated by a new line character
+        String paths = Arrays.stream(values)
+                .map(f -> f.getAbsolutePath())
+                .collect(Collectors.joining("\n"));
+        setValue(paths);
+    }
+
+    /**
+     * Gets the Upload button of the widget.
+     * The button is only rendered in advanced mode.
+     * @return the widget's upload button
+     */
+    public WebElement getAdvancedUploadButton() {
+        return findElement(By.cssSelector("button.ui-fileupload-upload"));
+    }
+
+    /**
+     * Gets the Cancel button of the widget.
+     * The button is only rendered in advanced mode.
+     * @return the widget's cancel button
+     */
+    public WebElement getAdvancedCancelButton() {
+        return findElement(By.cssSelector("button.ui-fileupload-cancel"));
+    }
+
+    /**
+     * Waits until all selected files are uploaded.
+     * This only works in advanced mode.
+     */
+    public void waitAdvancedUntilAllFilesAreUploaded() {
+        ConfigProvider config = ConfigProvider.getInstance();
+        WebDriver driver = WebDriverProvider.get();
+
+        // todo: Should there be a separate configuration entry for file upload?
```

Comment:
Why not add this code to `PrimeExpectedConditions` class instead?

Reply:
I looked for a method that waits until all Javascript activity on client side is completed. But I did not find such a method. This method `waitAdvancedUntilAllFilesAreUploaded` only works for file upload widgets in advanced mode. From my point of view I would not move this code into `PrimeExpectedConditions`. But maybe there is a global solution to wait until all client side activity is completed, something like `PrimeSelenium#guardAjax()`.

Reply:
One idea may be using https://github.com/lightbody/browsermob-proxy#http-request-manipulation to monitor whether there is still a active request.

Reply:
@christophs78 looks promising. I suggest to evaluate this in a separate step (new issue?)

Reply:
Yes. Opening a new issue for this is a good idea. As always way more available options than available time. ;-) Let´s look what the next weeks/months bring. :-)
Overall i´m really happy with the progress we make on the integration-test-side. A huge thankyou for your contribution. It´s IMO the only way to provide a stable product and allow for continues safe refactoring.